File: unreal/NeshamaSDK/Source/NeshamaSDK/Private/SoulEngine/Emotion/SentimentAnalyzer.cpp
#include "SoulEngine/Emotion/SentimentAnalyzer.h"

static float GetEmoValue(const TMap<FString, float>& Dict, const FString& Key)
{
	const float* Val = Dict.Find(Key);
	return Val ? *Val : 0.0f;
}

FResponseHint FHintGenerator::Generate(
	const TMap<FString, float>& Emotions,
	const FCompositeEmotionResult& Composite)
{
	FResponseHint Hint;
	Hint.Tone = EResponseTone::Neutral;
	Hint.Urgency = EUrgency::Low;
	Hint.SuggestedActions.Add(ESuggestedAction::DialogueFriendly);
	Hint.Confidence = 0.0f;
	Hint.Reasoning = TEXT("No strong emotion detected");

	if (Emotions.Num() == 0) return Hint;

	float Anger = GetEmoValue(Emotions, TEXT("anger"));
	float Fear = GetEmoValue(Emotions, TEXT("fear"));
	float Joy = GetEmoValue(Emotions, TEXT("joy"));
	float Trust = GetEmoValue(Emotions, TEXT("trust"));
	float Sadness = GetEmoValue(Emotions, TEXT("sadness"));
	float Surprise = GetEmoValue(Emotions, TEXT("surprise"));
	float Disgust = GetEmoValue(Emotions, TEXT("disgust"));

	// Anger-based responses
	if (Anger > 0.6f)
	{
		Hint.Tone = EResponseTone::Hostile;
		Hint.Urgency = Anger > 0.8f ? EUrgency::High : EUrgency::Medium;
		Hint.SuggestedActions = { ESuggestedAction::DialogueHostile };
		if (Anger > 0.8f) Hint.SuggestedActions.Add(ESuggestedAction::QuestRefuse);
		Hint.Reasoning = FString::Printf(TEXT("High anger (%.2f)"), Anger);
	}
	else if (Anger > 0.3f)
	{
		Hint.Tone = EResponseTone::Angry;
		Hint.Urgency = EUrgency::Medium;
		Hint.SuggestedActions = { ESuggestedAction::DialogueHostile, ESuggestedAction::WithholdInfo };
		Hint.Reasoning = FString::Printf(TEXT("Moderate anger (%.2f)"), Anger);
	}
	// Fear-based responses
	else if (Fear > 0.6f)
	{
		Hint.Tone = EResponseTone::Fearful;
		Hint.Urgency = EUrgency::High;
		Hint.SuggestedActions = { ESuggestedAction::DialogueCautious, ESuggestedAction::Flee };
		Hint.Reasoning = FString::Printf(TEXT("High fear (%.2f)"), Fear);
	}
	else if (Fear > 0.3f)
	{
		Hint.Tone = EResponseTone::Nervous;
		Hint.Urgency = EUrgency::Medium;
		Hint.SuggestedActions = { ESuggestedAction::DialogueCautious };
		Hint.Reasoning = FString::Printf(TEXT("Moderate fear (%.2f)"), Fear);
	}
	// Joy-based responses
	else if (Joy > 0.5f)
	{
		Hint.Tone = EResponseTone::Joyful;
		Hint.Urgency = EUrgency::Low;
		Hint.SuggestedActions = { ESuggestedAction::DialogueFriendly, ESuggestedAction::Celebration };
		if (Trust > 0.4f)
		{
			Hint.SuggestedActions.Add(ESuggestedAction::ShareInfo);
			Hint.SuggestedActions.Add(ESuggestedAction::QuestOffer);
		}
		Hint.Reasoning = FString::Printf(TEXT("High joy (%.2f)"), Joy);
	}
	// Trust-based responses
	else if (Trust > 0.5f)
	{
		Hint.Tone = EResponseTone::Trusting;
		Hint.Urgency = EUrgency::Low;
		Hint.SuggestedActions = { ESuggestedAction::DialogueFriendly, ESuggestedAction::ShareInfo };
		if (Composite.Name == TEXT("love"))
			Hint.SuggestedActions.Add(ESuggestedAction::QuestOffer);
		Hint.Reasoning = FString::Printf(TEXT("High trust (%.2f)"), Trust);
	}
	// Sadness-based responses
	else if (Sadness > 0.4f)
	{
		Hint.Tone = EResponseTone::Sad;
		Hint.Urgency = EUrgency::Low;
		Hint.SuggestedActions = { ESuggestedAction::Consolation, ESuggestedAction::DialogueCautious };
		Hint.Reasoning = FString::Printf(TEXT("Sadness detected (%.2f)"), Sadness);
	}
	// Surprise-based responses
	else if (Surprise > 0.5f)
	{
		Hint.Tone = EResponseTone::Surprised;
		Hint.Urgency = EUrgency::Medium;
		Hint.SuggestedActions = { ESuggestedAction::DialogueCautious };
		Hint.Reasoning = FString::Printf(TEXT("Surprise detected (%.2f)"), Surprise);
	}
	// Disgust-based responses
	else if (Disgust > 0.4f)
	{
		Hint.Tone = EResponseTone::Hostile;
		Hint.Urgency = EUrgency::Medium;
		Hint.SuggestedActions = { ESuggestedAction::DialogueHostile, ESuggestedAction::QuestRefuse };
		Hint.Reasoning = FString::Printf(TEXT("Disgust detected (%.2f)"), Disgust);
	}

	// Composite emotion overrides (higher priority)
	if (Composite.Intensity > 0.5f)
	{
		if (Composite.Name == TEXT("gratitude"))
		{
			Hint.Tone = EResponseTone::Grateful;
			Hint.SuggestedActions = { ESuggestedAction::GiveGift, ESuggestedAction::ShareInfo };
			Hint.Reasoning = TEXT("Gratitude composite emotion");
		}
		else if (Composite.Name == TEXT("pride"))
		{
			Hint.Tone = EResponseTone::Proud;
			Hint.SuggestedActions = { ESuggestedAction::Celebration, ESuggestedAction::QuestOffer };
			Hint.Reasoning = TEXT("Pride composite emotion");
		}
		else if (Composite.Name == TEXT("love"))
		{
			Hint.Tone = EResponseTone::Trusting;
			Hint.SuggestedActions = { ESuggestedAction::ShareInfo, ESuggestedAction::QuestOffer };
			Hint.Reasoning = TEXT("Love composite emotion");
		}
	}

	// Calculate confidence
	int32 NumEmotions = Emotions.Num();
	if (NumEmotions <= 2) Hint.Confidence = 0.9f;
	else
	{
		float MaxIntensity = 0.0f, MinIntensity = TNumericLimits<float>::Max();
		for (const auto& KV : Emotions)
		{
			MaxIntensity = FMath::Max(MaxIntensity, KV.Value);
			MinIntensity = FMath::Min(MinIntensity, KV.Value);
		}
		if (MaxIntensity - MinIntensity > 0.5f) Hint.Confidence = 0.85f;
		else if (NumEmotions >= 4) Hint.Confidence = 0.6f;
		else Hint.Confidence = 0.75f;
	}

	return Hint;
}

File: unity/NeshamaSDK/Runtime/SoulEngine/Emotion/SentimentAnalyzer.cs
using System;
using System.Collections.Generic;

namespace Neshama.SoulEngine.Emotion
{
    /// <summary>
    /// Generates response hints from emotion state.
    /// Ported from Python fast_path.py EmotionFastPath._generate_hint() and _infer_response().
    /// </summary>
    public static class HintGenerator
    {
        /// <summary>
        /// Generate a response hint from emotion state and composite emotion.
        /// </summary>
        public static ResponseHint Generate(
            Dictionary<string, float> emotions,
            CompositeEmotionResult composite)
        {
            var (tone, urgency, actions, reasoning) = InferResponse(emotions, composite);
            float confidence = CalculateConfidence(emotions);

            return new ResponseHint
            {
                tone = tone,
                urgency = urgency,
                suggestedActions = actions.ToArray(),
                confidence = confidence,
                reasoning = reasoning,
            };
        }

        /// <summary>
        /// Infer response parameters from emotion state.
        /// Ported from Python EmotionFastPath._infer_response().
        /// </summary>
        private static (ResponseTone, Urgency, List<SuggestedAction>, string) InferResponse(
            Dictionary<string, float> emotions,
            CompositeEmotionResult composite)
        {
            ResponseTone tone = ResponseTone.Neutral;
            Urgency urgency = Urgency.Low;
            var actions = new List<SuggestedAction> { SuggestedAction.DialogueFriendly };
            string reasoning = "No strong emotion detected";

            if (emotions == null || emotions.Count == 0)
                return (tone, urgency, actions, reasoning);

            float anger = GetValue(emotions, "anger");
            float fear = GetValue(emotions, "fear");
            float joy = GetValue(emotions, "joy");
            float trust = GetValue(emotions, "trust");
            float sadness = GetValue(emotions, "sadness");
            float surprise = GetValue(emotions, "surprise");
            float disgust = GetValue(emotions, "disgust");

            // Anger-based responses
            if (anger > 0.6f)
            {
                tone = ResponseTone.Hostile;
                urgency = anger > 0.8f ? Urgency.High : Urgency.Medium;
                actions = new List<SuggestedAction> { SuggestedAction.DialogueHostile };
                if (anger > 0.8f) actions.Add(SuggestedAction.QuestRefuse);
                reasoning = $"High anger ({anger:F2})";
            }
            else if (anger > 0.3f)
            {
                tone = ResponseTone.Angry;
                urgency = Urgency.Medium;
                actions = new List<SuggestedAction> { SuggestedAction.DialogueHostile, SuggestedAction.WithholdInfo };
                reasoning = $"Moderate anger ({anger:F2})";
            }
            // Fear-based responses
            else if (fear > 0.6f)
            {
                tone = ResponseTone.Fearful;
                urgency = Urgency.High;
                actions = new List<SuggestedAction> { SuggestedAction.DialogueCautious, SuggestedAction.Flee };
                reasoning = $"High fear ({fear:F2})";
            }
            else if (fear > 0.3f)
            {
                tone = ResponseTone.Nervous;
                urgency = Urgency.Medium;
                actions = new List<SuggestedAction> { SuggestedAction.DialogueCautious };
                reasoning = $"Moderate fear ({fear:F2})";
            }
            // Joy-based responses
            else if (joy > 0.5f)
            {
                tone = ResponseTone.Joyful;
                urgency = Urgency.Low;
                actions = new List<SuggestedAction> { SuggestedAction.DialogueFriendly, SuggestedAction.Celebration };
                if (trust > 0.4f)
                {
                    actions.Add(SuggestedAction.ShareInfo);
                    actions.Add(SuggestedAction.QuestOffer);
                }
                reasoning = $"High joy ({joy:F2})";
            }
            // Trust-based responses
            else if (trust > 0.5f)
            {
                tone = ResponseTone.Trusting;
                urgency = Urgency.Low;
                actions = new List<SuggestedAction> { SuggestedAction.DialogueFriendly, SuggestedAction.ShareInfo };
                if (composite.name == "love")
                    actions.Add(SuggestedAction.QuestOffer);
                reasoning = $"High trust ({trust:F2})";
            }
            // Sadness-based responses
            else if (sadness > 0.4f)
            {
                tone = ResponseTone.Sad;
                urgency = Urgency.Low;
                actions = new List<SuggestedAction> { SuggestedAction.Consolation, SuggestedAction.DialogueCautious };
                reasoning = $"Sadness detected ({sadness:F2})";
            }
            // Surprise-based responses
            else if (surprise > 0.5f)
            {
                tone = ResponseTone.Surprised;
                urgency = Urgency.Medium;
                actions = new List<SuggestedAction> { SuggestedAction.DialogueCautious };
                reasoning = $"Surprise detected ({surprise:F2})";
            }
            // Disgust-based responses
            else if (disgust > 0.4f)
            {
                tone = ResponseTone.Hostile;
                urgency = Urgency.Medium;
                actions = new List<SuggestedAction> { SuggestedAction.DialogueHostile, SuggestedAction.QuestRefuse };
                reasoning = $"Disgust detected ({disgust:F2})";
            }

            // Composite emotion overrides (higher priority)
            if (composite.intensity > 0.5f)
            {
                if (composite.name == "gratitude")
                {
                    tone = ResponseTone.Grateful;
                    actions = new List<SuggestedAction> { SuggestedAction.GiveGift, SuggestedAction.ShareInfo };
                    reasoning = "Gratitude composite emotion";
                }
                else if (composite.name == "pride")
                {
                    tone = ResponseTone.Proud;
                    actions = new List<SuggestedAction> { SuggestedAction.Celebration, SuggestedAction.QuestOffer };
                    reasoning = "Pride composite emotion";
                }
                else if (composite.name == "love")
                {
                    tone = ResponseTone.Trusting;
                    actions = new List<SuggestedAction> { SuggestedAction.ShareInfo, SuggestedAction.QuestOffer };
                    reasoning = "Love composite emotion";
                }
            }

            return (tone, urgency, actions, reasoning);
        }

        /// <summary>
        /// Calculate confidence in the emotion interpretation.
        /// Ported from Python EmotionFastPath._calculate_confidence().
        /// </summary>
        private static float CalculateConfidence(Dictionary<string, float> emotions)
        {
            if (emotions == null || emotions.Count == 0) return 0.5f;

            int numEmotions = emotions.Count;
            if (numEmotions <= 2) return 0.9f;

            float maxIntensity = 0f, minIntensity = float.MaxValue;
            foreach (var kv in emotions)
            {
                if (kv.Value > maxIntensity) maxIntensity = kv.Value;
                if (kv.Value < minIntensity) minIntensity = kv.Value;
            }

            // Clear dominance = high confidence
            if (maxIntensity - minIntensity > 0.5f) return 0.85f;

            // Many weak emotions = lower confidence
            if (numEmotions >= 4) return 0.6f;

            return 0.75f;
        }

        private static float GetValue(Dictionary<string, float> dict, string key)
        {
            float val;
            return dict != null && dict.TryGetValue(key, out val) ? val : 0f;
        }
    }
}
